tests/dataimportcron_test.go

```diff
@@ -0,0 +1,219 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+	cdiclientset "kubevirt.io/containerized-data-importer/pkg/client/clientset/versioned"
+	"kubevirt.io/containerized-data-importer/pkg/controller"
+	"kubevirt.io/containerized-data-importer/tests/framework"
+	"kubevirt.io/containerized-data-importer/tests/utils"
+)
+
+const (
+	dataImportCronTimeout = 90 * time.Second
+)
+
+var _ = Describe("DataImportCron", func() {
+	const (
+		scheduleEveryMinute = "* * * * *"
+		scheduleOnceAYear   = "0 0 1 1 *"
+	)
+	var (
+		f                   = framework.NewFramework(namespacePrefix)
+		registryPullNode    = cdiv1.RegistryPullNode
+		trustedRegistryURL  = func() string { return fmt.Sprintf(utils.TrustedRegistryURL, f.DockerPrefix, f.DockerTag) }
+		externalRegistryURL = "docker://quay.io/kubevirt/fedora-cloud-registry-disk-demo:latest"
+		cron                *cdiv1.DataImportCron
+		err                 error
+	)
+
+	AfterEach(func() {
+		if cron != nil {
+			By("Delete cron")
+			err = DeleteDataImportCron(f.CdiClient, cron.Namespace, cron.Name)
+			Expect(err).ToNot(HaveOccurred())
+		}
+	})
+
+	table.DescribeTable("should", func(schedule string, repeat int) {
+		var url string
+
+		if utils.IsOpenshift(f.K8sClient) {
+			url = externalRegistryURL
+		} else {
+			url = trustedRegistryURL()
+			err = utils.AddInsecureRegistry(f.CrClient, url)
+			Expect(err).To(BeNil())
+
+			hasInsecReg, err := utils.HasInsecureRegistry(f.CrClient, url)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(hasInsecReg).To(BeTrue())
+			defer utils.RemoveInsecureRegistry(f.CrClient, url)
+		}
+
+		cron = NewDataImportCron("cron-test", "5Gi", schedule, "datasource-test", cdiv1.DataVolumeSourceRegistry{URL: &url, PullMethod: &registryPullNode})
+		By(fmt.Sprintf("Create new DataImportCron %s", url))
+		cron, err = CreateDataImportCronFromDefinition(f.CdiClient, f.Namespace.Name, cron)
+		Expect(err).ToNot(HaveOccurred())
+
+		if schedule == scheduleEveryMinute {
+			var cronjob *v1beta1.CronJob
+			By("Verify cronjob was created")
+			Eventually(func() bool {
+				cronjob, err = f.K8sClient.BatchV1beta1().CronJobs(f.CdiInstallNs).Get(context.TODO(), controller.GetCronJobName(cron), metav1.GetOptions{})
+				return err == nil && len(cronjob.Status.Active) > 0
+			}, dataImportCronTimeout, pollingInterval).Should(BeTrue())
+			Expect(err).ToNot(HaveOccurred())
+
+			By(fmt.Sprintf("Find pod of job %s", cronjob.Status.Active[0].Name))
+			Eventually(func() bool {
+				_, err = utils.FindPodByPrefix(f.K8sClient, f.CdiInstallNs, cronjob.Status.Active[0].Name, "")
+				return err == nil
+			}, dataImportCronTimeout, pollingInterval).Should(BeTrue())
+		}
+
+		var lastImportDv, currentImportDv string
+		for i := 0; i < repeat; i++ {
+			if repeat > 1 {
+				// Emulate source update
+				digest := []string{"sha256:0001112223334444", "sha256:0123456789abcdef", "sha256:aaabbbcccdddeeee"}[i]
```

Comment:
clearly these are fake shas, so I'm wondering why import succeeds.  Shouldn't an import be bound to a specific sha?

Reply:
The 1st test case actually uses the cronjob (which uses skopeo inspect to read the source image digest sha, and update the DataImportCron DesiredDigest annotation accordingly)
The 2nd & 3rd test cases simply emulate the cronjob app to test source image updates. So were are just updating the sha which is enough to test all but the cronjob.

Reply:
I am thinking that Michaels point is that we should be grabbing the exact sha, that is in the desired, this way we can guarantee we are getting the correct one. What if somehow in between us scheduling the job, and the job completing and triggering the controller to create the DV, a new version becomes available in the registry, and the sha is different. We would want the desired sha to actually be imported instead of :latest which may not be the sha we wanted.

Reply:
(1) seems reasonable and I may fix it, but
(2) afaiu all we are interested is always importing the latest update, so in the rare case described  that's what happens but we'll  have an entry with the old sha and the new DV name will be based on this old sha. However it will be imported again since the source sha updated.
(3) the testing method suggested here won't work, and we'll need to find a smarter way to do it...
(4) "Perfection is the enemy of progress." -Winston Churchill

Reply:
No we are interested in the exact sha being in the DV that the datasource points to. There is a reason we keep some of the older ones around, so we can go back to an exact version indicated by the sha. If that is not always the case there is no point in keeping older ones around.
For the test, you are right the current approach won't work, but we could figure out the shas we are interested in ahead of time by basically running the script in the job, and collecting the sha from that. Don't we return it in the exit code? We could do that for some of the older versions of that image in quay, collect the shas from there, and then use that in the test. It should then succeed.

Reply:
Convinced. Fixing.